### ISSUE TITLE
cmake can't find sac2c

### ISSUE THREAD

**Issue by bernecky** [SacBase/Dislin#2]
I am trying to install dislin on Ubuntu 20.10, and "cmake .." can't find sac2c:

```
sac@boa:~/sac/BASE/dislin/build$ cmake ..
-- The C compiler identification is GNU 10.2.0
...
-- Detecting CXX compile features - done
CMake Error at cmake-common/check-sac2c.cmake:9 (MESSAGE):
  Could not locate the sac2c binary, exiting...
Call Stack (most recent call first):
  CMakeLists.txt:17 (INCLUDE)

-- Configuring incomplete, errors occurred!
See also "/home/sac/sac/BASE/dislin/build/CMakeFiles/CMakeOutput.log".
sac@boa:~/sac/BASE/dislin/build$ cmake --version
cmake version 3.16.3

```
sac2c is in $PATH, and it lives at /usr/local/bin/sac2c. Cmake's find-program docs say that it should
search the PATH environment variable.

1. I have not figured out how to make cmake find-program tell me WHERE it looks for sac2c. My attempts to use any debug/verbose options are sans fruit.
2. I have a slightly old version of cmake, but Ubuntu says it is the latest (it's not). However, the last update of dislin here was several years ago, so...
3. I did do submodule init; submodule update, so I think that part of the world is good.

If somebody could try a fresh "cmake .." of dislin,  please let us know the results of that, and also your version of cmake.

**Comment by ashinkarov** [comment]
Hmm, on my system it also doesn't pick up `sac2c` when it is located in a non-standard location which is on the PATH.  Probably we are missing some flag of sorts.

As a workaround for now, run cmake as follows: `cmake -DSAC2C_EXEC=/usr/local/bin/sac2c ..` that should do the right thing.  Meanwhile, @hv15 , any ideas what is missing?

**Comment by ashinkarov** [comment]
Ok, I figured out the problem, but  I am not sure what the solution is. @hv15 , any ideas?
The problem is that `SAC2C_EXEC` is a cached string.  It is fine if we set it externally, but we cannot (by some weird reason) change its value during the execution of the script.  Here is a reduced example:
```sh
/tmp|=> cd test 
test|=> cat CMakeLists.txt 
CMAKE_MINIMUM_REQUIRED (VERSION 3.3)
# SET (XXX   CACHE STRING "blah")
FIND_PROGRAM (XXX NAMES sac2c) 
MESSAGE ("XXX: ${XXX}")
test|=> which sac2c
/home/tema/local/bin/sac2c
test|=> sac2c
I am sac2c!
test|=> cd build; cmake ..
-- ...
XXX: /home/tema/local/bin/sac2c
-- Configuring done
test|=> cat CMakeLists.txt 
CMAKE_MINIMUM_REQUIRED (VERSION 3.3)
SET (XXX   CACHE STRING "blah")
FIND_PROGRAM (XXX NAMES sac2c) 
MESSAGE ("XXX: ${XXX}")
test|=> cd build; cmake ..
-- ...
XXX: 
-- Configuring done
```
WTF?

**Comment by bernecky** [comment]
I think we are all going down similar roads. 

Ah, commenting out this line in CMakeLists.txt lets
cmake find sac2c properly (as you can see from CMakeCache.txt after running "cmake ..":

`##SET (SAC2C_EXEC                     CACHE STRING "A path to sac2c compiler")`

It then fails a few lines later, but that is another kettle of fish.
I get the impression (and thought that I read it somewhere, but can't
confirm that at present) that cmake considers variables to be write-once. Hence, once set, new sets
will not overwrite its value. [That may only be true on a function-by-function basis. E.g., find-program's
first argument is treated as write-once.]

I am going to look into a few cmake things here:

1. Are all variables write-once? If not, what are the rules?
2. Why does the above SET put a value in the CMakeCache.txt, even though the lines does not specify a value?

**Comment by ashinkarov** [comment]
The first one is easy.  The answer is no:
```sh
build|=> cat ../CMakeLists.txt 
SET (XXX 1)
MESSAGE ("XXX: ${XXX}")
SET (XXX 2)
MESSAGE ("XXX: ${XXX}")
build|=> cmake ..
XXX: 1
XXX: 2
-- Configuring done
```
The second one: well, because of `CACHED`.  I think the reasoning behind using cached was: in case you call cmake again, the location of the sac2c would be already set.

**Comment by bernecky** [comment]
Re "look into", above: 
1. I find the cmake documentation imprecise and vague. E.g., the treatment of Variables sometimes says
it will "clear any existing value". I have no idea what "clear" means in this context.
Other times it mentions what happens to the CACHE, etc.

2. Changing the above ##SET to UNSET fixed things. The "SET" with no value looks to my unpracticed eye to be a cmake bug, in not reporting it as an error, since the docn describes the variable as the mandatory <VALUE>, rather than as the optional [<VALUE>].

I will check in that fix later.

**Comment by hv15** [comment]
I think I've spotted the problem, its coming from `check-sac2c.cmake` in [cmake-common](https://github.com/SacBase/cmake-common/). The issue is that because `SAC2C_EXEC` is not set (unless set externally) `find_program` goofs up. @ashinkarov you actually fixed this in https://github.com/SacBase/cmake-common/commit/1e05f51a966621c614970afe218ad46531d96efd.

@bernecky I would suggest you update the cmake-common submodule in the repo to the latest version. `git submodule update --recursive --remote` should do the trick.

**Comment by bernecky** [comment]
Thanks, Hans. I did not know about '--recursive or --remote", but they do look like things I want to add to my
standard submodule mantra.

I thought that my UNSET fix did the job, but I see now that it does the UNSET too late, so any explicit set
of SAC2C_EXEC in the cmake command line is evaporated. Sorry about that. I'll confirm that Artem's belated
(to me) fixes work, then close this issue.

**Comment by ashinkarov** [comment]
> @ashinkarov you actually fixed this in [SacBase/cmake-common@1e05f51](https://github.com/SacBase/cmake-common/commit/1e05f51a966621c614970afe218ad46531d96efd).

Ha, interesting.  Indeed, it fixes the problem, and I completely forgot that I made this change.
However, maybe this suggests that we should force submodule updates whenever we change cmake-common?  Not sure what would be the best way to achieve this...

**Comment by hv15** [comment]
@ashinkarov Not sure either, I do not think there is such a operation on github side of things... best case might be a git hook?

@bernecky no worries, `--remote` is relatively new I only found it today actually :sweat_smile:

**Comment by bernecky** [comment]
A fresh install of dislin (The real one, from dislin.com, not the SAC one, which I think should be renamed to SACDISLIN or something) installs to /usr/local by default, so cmake does not find dislin.h. No include folder visible there.

Do we have a preferred way to fix this? I can do the legwork, if we can agree on direction.

**Comment by hv15** [comment]
bizarre, there are multiple web pages that host dislin, I guess Helmut Michels got a bit overenthusiastic (there is the original site of the project http://www.mps.mpg.de/dislin, the general site www.dislin.de (and .com), _and_ the company www.dislin-software.de). The last two seem to be kept up-to-date...

Looking at the DEB distributable, there is no `include` directory. `dislin.h` is located at the root of `/usr/local/dislin`, which is just non-standard... best way I think to get this working is to modify the `cmake/check-dislin.cmake` file by changing the `find_program` macro to include `HINTS /usr/local/dislin`. Additionally setting the `CMAKE_REQUIRED_INCLUDES` to include the path `/usr/local/dislin` so that `check_include_files` can find the header file.

**Comment by bernecky** [comment]
Commit b6efaff4ec61c11844a50d206290d019fee02f88 claims to fix this stuff. Sort of.
dislin installs in non-standard places. I may have convinced the dislin author to add lib and include folders,
containing symlinks, to his package. That would make the file structure more "standard", yet leave it compatible
with how it is now.

I had to put in two hard-coded paths to /usr/local/dislin to get cmake to work properly. At that point, make more
or less worked. See next issue re "make -j". I did a "make test", but there are no steenking "make test"s.
